realm/realm-library/src/main/java/io/realm/internal/android/JsonUtils.java

```diff
@@ -37,13 +41,26 @@
      */
     public static Date stringToDate(String date) {
         if (date == null || date.length() == 0) return null;
+
+        // Check for JSON date
         Matcher matcher = jsonDate.matcher(date);
         if (matcher.find()) {
             String dateMatch = matcher.group(1);
             return new Date(Long.parseLong(dateMatch));
-        } else {
+        }
+
+        // Check for millisecond based date
+        if(numericOnly.matcher(date).matches()) {
             return new Date(Long.parseLong(date));
         }
+
+        // Try for ISO8601 date
+        try {
+            parsePosition.setIndex(0); // reset the position each time
+            return ISO8601Utils.parse(date, parsePosition);
+        } catch (ParseException ignored) {
+            return null;
```

Comment:
@cmelchior The previous version threw a `NumberFormatException`. In this case, that could still happen and the docs still have that in the throws documentation. If we are to throw a `ParseException` then we will have to add the throws declaration to the method signature. That could cause a ripple of changes because `ParseException` is a typical `Exception` while `NumberFormatException` is a `RunTimeException`. Alternatively, we can catch the parse exception and throw a different `RuntimeException` (one of our own, etc). What would you like to do here? Thoughts?